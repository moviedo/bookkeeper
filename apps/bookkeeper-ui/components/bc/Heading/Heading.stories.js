import { Heading } from "./Heading";

export default {
  title: "base components / bc-heading",
  argTypes: {
    type: {
      control: "select",
      options: Object.values(Heading),
      description: "h1-h6",
      defaultValue: "h1",
    },
    default: {
      control: "text",
      description: "default slot",
      defaultValue: "Heading 1",
    },
    subtitle: {
      control: "text",
      description: "subtitle slot that's automatically wrapped in hgroup",
    },
    group: {
      control: "text",
      description: "group slot that's automatically wrapped in hgroup",
    },
  },
};

export const Typography = () => ({
  template: `
    <div>
      <bc-heading type="h1">Heading 1</bc-heading>
      <bc-heading type="h2">Heading 2</bc-heading>
      <bc-heading type="h3">Heading 3</bc-heading>
      <bc-heading type="h4">Heading 4</bc-heading>
      <bc-heading type="h5">Heading 5</bc-heading>
      <bc-heading type="h6">Heading 6</bc-heading>
    </div>
  `,
});

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes).filter(
    (a) => !["default", "subtitle", "group"].includes(a)
  ),
  template: `
    <bc-heading v-bind="$props">
      ${args.default}
      <template v-if="${"subtitle" in args}" #subtitle>
        ${args.subtitle}
      </template>
      <template v-if="${"group" in args}" #group>
        ${args.group}
      </template>
    </bc-heading>`,
});

export const WithSubtitle = Template.bind({});
WithSubtitle.args = {
  subtitle: "Some subtitle",
};

export const GroupedHeading = Template.bind({});
GroupedHeading.args = {
  group: '<bc-heading type="h2">Heading 2</bc-heading>',
};

export const WithCustomAttr = Template.bind({});
WithCustomAttr.args = {
  group: '<bc-heading type="h2" style="color:red;">Heading 2</bc-heading>',
};
