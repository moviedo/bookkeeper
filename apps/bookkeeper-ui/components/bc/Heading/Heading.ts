import Vue, { PropOptions, VNode } from "vue";

export enum Heading {
  h1 = "h1",
  h2 = "h2",
  h3 = "h3",
  h4 = "h4",
  h5 = "h5",
  h6 = "h6",
}

export default Vue.extend({
  name: "BcHeading",
  props: {
    type: {
      type: Object,
      default: Heading.h1,
    } as PropOptions<Heading>,
  },
  render(createElement): VNode {
    const headingEl: Heading = this.$props.type;
    const heading = createElement(
      headingEl,
      {
        class: [
          getTextSize(headingEl),
          getTestWeight(headingEl),
          "text-slate-800",
          this.$slots.subtitle && !this.$slots.group ? "mb-2" : "",
        ],
      },
      this.$slots.default
    );

    if (this.$slots.subtitle && !this.$slots.group) {
      return createElement("hgroup", [
        heading,
        createElement(
          "p",
          { class: "text-base font-medium" },
          this.$slots.subtitle
        ),
      ]);
    } else if (this.$slots.group) {
      return createElement("hgroup", [heading, this.$slots.group]);
    }

    return heading;
  },
});

function getTextSize(headingEl: Heading): string {
  switch (headingEl) {
    case Heading.h5:
      return "font-bold";
    case Heading.h3:
      return "font-medium";
    default:
      return "font-normal";
  }
}

function getTestWeight(headingEl: Heading): string {
  switch (headingEl) {
    case Heading.h6:
    case Heading.h5:
      return "text-xl";
    case Heading.h4:
      return "text-2xl";
    case Heading.h3:
      return "text-3xl";
    case Heading.h2:
      return "text-4xl";
    default:
      return "text-5xl";
  }
}
