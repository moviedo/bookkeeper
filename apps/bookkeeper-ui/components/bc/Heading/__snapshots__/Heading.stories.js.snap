// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base components / bc-heading GroupedHeading smoke-test 2`] = `
<hgroup>
  <h1 class="font-normal text-5xl text-slate-800">
    Heading 1
  </h1>
  <h2 class="font-normal text-4xl text-slate-800">
    Heading 2
  </h2>
</hgroup>
`;

exports[`base components / bc-heading Typography smoke-test 2`] = `
<div>
  <h1 class="font-normal text-5xl text-slate-800">
    Heading 1
  </h1>
  <h2 class="font-normal text-4xl text-slate-800">
    Heading 2
  </h2>
  <h3 class="font-medium text-3xl text-slate-800">
    Heading 3
  </h3>
  <h4 class="font-normal text-2xl text-slate-800">
    Heading 4
  </h4>
  <h5 class="font-bold text-xl text-slate-800">
    Heading 5
  </h5>
  <h6 class="font-normal text-xl text-slate-800">
    Heading 6
  </h6>
</div>
`;

exports[`base components / bc-heading WithSubtitle smoke-test 2`] = `
<hgroup>
  <h1 class="font-normal text-5xl text-slate-800 mb-2">
    Heading 1
  </h1>
  <p class="text-base font-medium">
    Some subtitle
  </p>
</hgroup>
`;
